src/common/mpu6050.rs

```diff
@@ -0,0 +1,348 @@
+    // temperature: f64,
+    error: Option<Arc<anyhow::Error>>,
+}
+
+impl MPU6050State {
+    fn new() -> Self {
+        MPU6050State {
+            angular_velocity: Vector3::new(),
+            linear_acceleration: Vector3::new(),
+            // temperature: 0.0,
+            error: None,
+        }
+    }
+
+    fn get_angular_velocity(&self) -> Vector3 {
+        self.angular_velocity
+    }
+
+    fn get_linear_acceleration(&self) -> Vector3 {
+        self.linear_acceleration
+    }
+
+    fn set_error(&mut self, err: Option<Arc<anyhow::Error>>) {
+        self.error = err;
+    }
+
+    fn set_linear_acceleration_from_reading(&mut self, reading: &[u8; 14]) {
+        let mut slice_copy = vec![0; 6];
+        slice_copy.clone_from_slice(&(reading[0..6]));
+        let mut rdr = Cursor::new(slice_copy);
+        let unscaled_x = rdr.read_i16::<BigEndian>().unwrap();
+        let unscaled_y = rdr.read_i16::<BigEndian>().unwrap();
+        let unscaled_z = rdr.read_i16::<BigEndian>().unwrap();
+
+        let max_acceleration: f64 = 2.0 * 9.81 * 1000.0;
+
+        let x = f64::from(unscaled_x) * max_acceleration / 32768.0;
+        let y = f64::from(unscaled_y) * max_acceleration / 32768.0;
+        let z = f64::from(unscaled_z) * max_acceleration / 32768.0;
+        self.linear_acceleration = Vector3 { x, y, z };
+    }
+
+    fn set_angular_velocity_from_reading(&mut self, reading: &[u8; 14]) {
+        let mut slice_copy = vec![0; 6];
+        slice_copy.clone_from_slice(&(reading[8..14]));
+        let mut rdr = Cursor::new(slice_copy);
+        let unscaled_x = rdr.read_u16::<BigEndian>().unwrap();
+        let unscaled_y = rdr.read_u16::<BigEndian>().unwrap();
+        let unscaled_z = rdr.read_u16::<BigEndian>().unwrap();
+        let x = f64::from(unscaled_x) * 250.0 / 32768.0;
+        let y = f64::from(unscaled_y) * 250.0 / 32768.0;
+        let z = f64::from(unscaled_z) * 250.0 / 32768.0;
+        self.angular_velocity = Vector3 { x, y, z };
+    }
+}
+
+pub struct MPU6050 {
+    state: Arc<Mutex<MPU6050State>>,
+    i2c_handle: I2cHandleType,
+    i2c_address: u8,
+    canceller: Sender<bool>,
+}
+
+impl MPU6050 {
+    pub fn new(mut i2c_handle: I2cHandleType, i2c_address: u8) -> anyhow::Result<Self> {
+        let bytes: [u8; 2] = [107, 0];
+        i2c_handle.write_i2c(i2c_address, &bytes)?;
+        let i2c_address_copy = i2c_address;
+        let raw_state = MPU6050State::new();
+        let (canceller, rx) = mpsc::channel();
+        let state = Arc::new(Mutex::new(raw_state));
+        // reference copies for sending memory into thread
+        let mut i2c_handle_copy = Arc::clone(&i2c_handle);
+        let state_copy = Arc::clone(&state);
+        // start a polling thread that reads from the MPU every millisecond and mutates state.
+        // This allows multi-read access to the state for the functions satisfying the
+        // Movement Sensor API
+        thread::spawn(move || loop {
```

Comment:
we can't spawn a thread here on the esp32 platform we don't have much control on priority/stack allocation/core allocation and it could preempt the thread running the executor slowing down IO or worse deadlocking.

Can we just poll on demand for now? (eg read i2c when a grpc request comes in?) It will be slower but given the high latency I don't think it would be significant.

Ultimately we could spwan a future on the local executor